epochStart/shardchain/trigger.go

```diff
@@ -181,14 +190,56 @@ func NewEpochStartTrigger(args *ArgsShardEpochStartTrigger) (*trigger, error) {
 		appStatusHandler:            &statusHandler.NilStatusHandler{},
 	}
 
-	newTrigger.headersPool.RegisterHandler(newTrigger.receivedMetaBlock)
+	t.headersPool.RegisterHandler(t.receivedMetaBlock)
 
-	err := newTrigger.saveState(newTrigger.triggerStateKey)
+	err := t.saveState(t.triggerStateKey)
 	if err != nil {
 		return nil, err
 	}
 
-	return newTrigger, nil
+	t.mapMisingMiniblocks = make(map[string]*block.MiniBlock)
+	go t.requestMissingMiniblocks()
+
+	return t, nil
+}
+
+func (t *trigger) requestMissingMiniblocks() {
+	for {
+		time.Sleep(sleepTime)
+
+		t.mutMissingMiniblocks.RLock()
+		if len(t.mapMisingMiniblocks) == 0 {
+			t.mutMissingMiniblocks.RUnlock()
+			continue
+		}
+
+		missingMiniblocks := make([][]byte, len(t.mapMisingMiniblocks))
+		for hash := range t.mapMisingMiniblocks {
+			missingMiniblocks = append(missingMiniblocks, []byte(hash))
+			log.Debug("trigger.requestMissingMiniblocks", "hash", []byte(hash))
+		}
+		t.mutMissingMiniblocks.RUnlock()
+
+		go t.requestHandler.RequestMiniBlocks(core.MetachainShardId, missingMiniblocks)
+		time.Sleep(waitTime)
+		t.updateMissingMiniblocks()
+	}
+}
+
+func (t *trigger) updateMissingMiniblocks() {
+	t.mutMissingMiniblocks.Lock()
+	defer t.mutMissingMiniblocks.Unlock()
+
+	for hash := range t.mapMisingMiniblocks {
+		if t.miniBlocksPool.Has([]byte(hash)) {
```

Comment:
elements from the mapMissingMiniblocks should be deleted once the epoch has changed and was processed. No need to keep them. this can create out of memory problems.

Reply:
Actually the epoch could not be changed if there are more elements in this map. Anyway I would add a forced clean when epoch changed and processed